code/cabal-audit/src/Distribution/Audit.hs

```diff
@@ -0,0 +1,246 @@
+module Distribution.Audit (auditMain, buildAdvisories, AuditConfig (..), AuditException (..), Output (..)) where
+
+import Colourista.Pure (blue, bold, formatWith, green, red, yellow)
+import Control.Exception (Exception (displayException), SomeException (SomeException), catch, throwIO)
+import Control.Monad (when)
+import Data.Aeson (KeyValue ((.=)), Value, object)
+import Data.Aeson qualified as Aeson
+import Data.Coerce (coerce)
+import Data.Foldable (for_)
+import Data.Functor.Identity (Identity (runIdentity))
+import Data.List qualified as List
+import Data.Map qualified as M
+import Data.String (IsString (fromString))
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Text.IO qualified as T
+import Distribution.Client.NixStyleOptions (NixStyleFlags, defaultNixStyleFlags)
+import Distribution.Client.ProjectConfig (ProjectConfig)
+import Distribution.Client.ProjectOrchestration
+  ( CurrentCommand (OtherCommand)
+  , ProjectBaseContext (ProjectBaseContext, cabalDirLayout, distDirLayout, localPackages, projectConfig)
+  , commandLineFlagsToProjectConfig
+  , establishProjectBaseContext
+  )
+import Distribution.Client.ProjectPlanning (rebuildInstallPlan)
+import Distribution.Client.Setup (defaultGlobalFlags)
+import Distribution.Types.PackageName (PackageName, unPackageName)
+import Distribution.Verbosity qualified as Verbosity
+import Distribution.Version (Version, versionNumbers)
+import GHC.Generics (Generic)
+import Options.Applicative
+import Security.Advisories (Advisory (..), Keyword (..), ParseAdvisoryError, printHsecId)
+import Security.Advisories.Cabal (ElaboratedPackageInfoAdvised, ElaboratedPackageInfoWith (MkElaboratedPackageInfoWith, elaboratedPackageVersion, packageAdvisories), matchAdvisoriesForPlan)
+import Security.Advisories.Convert.OSV qualified as OSV
+import Security.Advisories.Filesystem (listAdvisories)
+import System.Exit (exitFailure)
+import System.IO.Temp (withSystemTempDirectory)
+import System.Process (callProcess)
+import Validation (validation)
+
+data AuditException
+  = -- | parsing the advisory database failed
+    ListAdvisoryValidationError {originalFilePath :: FilePath, parseError :: [ParseAdvisoryError]}
+  | -- | to rethrow exceptions thrown by cabal during plan elaboration
+    CabalException {reason :: String, cabalException :: SomeException}
+  deriving stock (Show, Generic)
+
+instance Exception AuditException where
+  displayException = \case
+    ListAdvisoryValidationError dir errs ->
+      mconcat
+        [ "Listing the advisories in directory "
+        , dir
+        , " failed with: \n"
+        , show errs
+        ]
+    CabalException ctx (SomeException ex) ->
+      "cabal failed while "
+        <> ctx
+        <> ":\n"
+        <> displayException ex
+
+-- | the type of output that is chosen for the command
+data Output
+  = -- | write humand readable to stdout
+    HumanReadable
+  | -- | write as Osv format to the specified file
+    Osv {osvPath :: FilePath}
```

Comment:
Why not to stdout? I suggest to make this `Maybe FilePath`.

Reply:
I have thought about that, but reverted back to just outputting to file as it simplifies the parser...

Reply:
But outputting to stdout is the unix way.

Reply:
"format" and "output" should be two different concerns. I should be able to write the "human readable" stuff to a file too.

Reply:
alright, I will change accordingly.

Reply:
@hasufell done 

- new option `-o` or `--to-file` that allows to specify a file path to write to (both human-readable and json) 
- option `--json` to output as json